docs/rfcs/20220804-ptensor/README.md

```diff
@@ -0,0 +1,268 @@
+# RFC: Parallel Tensor Dialect (PTensor)
+Frank Schlimbach, Ivan Butygin, Diptorup Deb
+
+## Summary
+We propose a solution which can provide a high-level dialect for tensors and (numpy-like) tensor-operations allowing passes to create parallelism for GPU, OpenMP but also distributed memory while following the compute-follows-data concept.
+
+In this document, we describe several dialects and passes to provide a good high-level view on the overall flow. We might want to split this into more than one document.
+
+## Motivation
+MLIR provides a few tensor dialects (like tensor and TOSA) and separate dialects/conversions for targeting devices and parallelism. An annotated type is needed to keep the required analysis in the compiler pipeline to an acceptable extent. Even moderately complicated flows, for example when device and host tensors exist at the same time, complicated analysis is needed to correctly implement compute-follows data. This is particualarly true when converting to distributed parallelism, such as MPI-based; this is difficult to get right with existing dialects if possible at all.
+
+Additionally, the current architecture in Plier does not have an explicit MLIR representation of arrays/tensors and so makes it very hard to re-use its tensor functionality without coming from python/numba. Currently conversions are even relying the Python runtime. A dialect for such numpy-like tensors and -operations can accept hints for targets (like devices, MPI etc) and enable its use in libraries independent of numba and Python. One example of such a use case is a distributed and JIT-compiler based numpy implementation.
+
+## Proposal
+We propose two new Dialects and one dialect extension:
+1. __PTensor__ dialect: providing a tensor type (ptensor) and operations on such ptensors
+2. __Dist__ dialect: dealing with the aspects of distributed data management, such as distributed memory alloction, GC and communication
+3. __intel_sycl.device_region__ operation in __intel_sycl__ dialect (RFC #291)
+
+
+Additionally we propose appropriate passes
+1. Converting __PTensor__ operations into dialects __Linalg__, __GPU__, __intel_sycl__ and __Dist__
+2. Converting __Dist__ operations into appropriate runtime calls
+3. Converting __intel-sycl.device_region__ to appropriate runtime calls.
+
+### ptensor Type
```

Comment:
We still need to decide (and document) if want value semantics like upstream `tensor` type, where each tensor is immutable, or reference semantics like upstream `memref` which allow inplace modifications. Most likely we need reference, to properly support numpy array assignment.

Reply:
Good point.

I can't find anything in the docs saying that the builtin 'tensor' is immutable. The tensor dialect actually has 'insert' and 'insert_slice' opterations.

In any case, my latest commit clarifies that ptensor is intended to be mutable and to lower to memref. This is aligned with Linalg.